### PR TITLE
Integrate Puppeteer to simulate user behaviour, create features for the probe to measure youtube search results in different scenario

### PR DESCRIPTION
**Pull request by 0x5eba**
- Use puppeteer for login instead of Youtube API
- Make a request every hours for 3 days and see if there is any time where the results doesn't show up
- Make a request to change the privacy: upload a video, wait to show up for everyone (all ips), change the privacy and then test the prober on it
- Same as above, but also with description, title etc.
- Make the average of the replay, and double to trigger an alert if the request take longer than it

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+
+Usage
+-----
+
+Client:
+```
+client [command]
+
+Commands:
```
Comment (harababurel) [inline README.md:35]: The client hangs without any message when no command is provided. Maybe `--help` should be the default behavior?

File: README.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    client upload-basic   Upload a random video and search it unitl all ips found it
```
Comment (harababurel) [inline README.md:36]: nit: typo

File: README.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    protocol buffer compiler.
+
+* `external`
+  * External dependencies for the project.  Currently only contains the
+    `node_modules` directory populated by npm using the standard methods.
+
+* `dev_container`
+  * Code for setting up and controlling the container used for building and
+    running the project.  The intention is to create a checked in and
+    reproducible state for the development environment, losening requirements on
```
Comment (harababurel) [inline README.md]: nit: typo

File: README.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    the host system used to do development.
+
+* `src/potato`
```
Comment (harababurel) [inline README.md]: Why potato? Is this a leftover from Brian's implementation?

File: client.js [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+        for(let a = 0; a < data.length; ++a){
+            if(data[a].success === false){
+                // TODO: Alert
+                console.error("ALERT", data[a])
+            }
+        }
+    })
+}
+
+async function mutilpleDaySearch(title, id_video) {
```
Comment (harababurel) [inline client.js]: typo

File: client.js [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+const https = require("https")
+const http = require("http")
+const express = require("express")
+const app = express()
+
+var ips = readJson("./ips.json")
+
+var socket_localhost = null
+
+async function _basicSearch(title, id_video) {
```
Comment (harababurel) [inline client.js]: `_basicSearch` and `basicSearch` have almost the same name but different behavior. It's hard to tell what each of them does without looking into the implementation. We should find a more suggestive name and add comments that describe behavior ([7.8 Method and function comments](https://google.github.io/styleguide/jsguide.html#jsdoc-method-and-function-comments)).

File: client.js [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+    var start = performance.now()
+    var promises = []
+    for (const [cell, ipscell] of Object.entries(ips)) {
+        for (let a = 0; a < ipscell.length - 1; ++a) {
+            let ip = "https://" + ipscell[a] + "/"
+            let country = cell
```
Comment (harababurel) [inline client.js]: `country` is a bit misleading in this scenario because of possible conflicts, e.g. there can be multiple cells in the same country.

File: client.js [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+                console.error("ALERT", data[a])
+            }
+        }
+    })
+}
+
+async function mutilpleDaySearch(title, id_video) {
+    var start = performance.now()
+    var interval = window.setInterval(() => {
+        if (performance.now() - start >= 86400000 * 3) { // 3 days
```
Comment (harababurel) [inline client.js]: Let's move this to a config file instead of hardcoding the value here.

File: client.js [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+            window.clearInterval(interval)
+        }
+        basicSearch(title, id_video)
+    }, 3600000) // hourly
```
Comment (harababurel) [inline client.js]: Same as above. Finer granularity would be better here (maybe once a minute?).

File: README.md [hunk 10]
```diff
@@ -0,0 +1,60 @@
+E2E search ml probing
```
Comment (harababurel) [inline README.md:1]: There's a README template that got lost along the way when this repo was created. Please merge the current readme with the extra parts which belong to the template and also add the two other files:

[README.md](https://gist.github.com/harababurel/fe30fe8a61f95a22c45aac7d5807750e)
[LICENSE](https://gist.github.com/harababurel/12bb872f75eceb148c3e93ab659cbe43)
[CONTRIBUTING.md](https://gist.github.com/harababurel/93f8cb3d4517c4c5aed0d221fa0cab91)

Also please note that every source file should contain a specific header (details in README).

File: client.js [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+}
+
+async function _searchMultipleParameters(title, id_video, description, is_public) {
+    var start = performance.now()
+    var promises = []
+    for (const [cell, ipscell] of Object.entries(ips)) {
+        for (let a = 0; a < ipscell.length - 1; ++a) {
```
Comment (harababurel) [inline client.js]: nit: `i`, `j`, `k` are more standard loop variable names (https://softwareengineering.stackexchange.com/questions/127934/any-standards-for-naming-variables-in-for-loops-instead-of-i-j-k)

File: client.js [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+                        })
+                })
+            )
+        }
+    }
+    return Promise.all(promises)
+}
+
+async function searchMultipleParameters(title, id_video, description, is_public) {
+    _searchMultipleParameters(title, id_video, description, is_public).then((data) => {
```
Comment (harababurel) [inline client.js]: Same as the previous comment about similarly named functions.

File: client.js [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+                            ok = true
+                        } else { // otherwise no
+                            ok = false
+                        }
+                    } else { // if is not the previous 2 cases it means is not correct
+                        ok = false
+                    }
+
+                    var t1 = performance.now()
+                    resolve([ok, parseInt(t1 - start_request), num_callback])
```
Comment (harababurel) [inline client.js]: This logic is a bit hard to follow. The `public_ok` and `private_ok` variable also add a bit of confusion over their meaning. I suggest setting `ok = true` initially and performing a series of checks that could possibly change its value to false. For example (excuse the C++-ized syntax):

```cpp
ok = true
if(!is_public && id_found) { ok = false; }
if(is_public && !id_found) { ok = false; }
if(is_public && !data.includes(description)) { ok = false; }
```
This can also eliminate some of the comments because the code itself reads more naturally.

File: client.js [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+        for(let a = 0; a < data.length; ++a){
+            if(data[a].success === false){
+                // TODO: Alert
+                console.error("ALERT", data[a])
+            }
+        }
+    })
+}
+
+async function searchIdAndFeatures(title, id_video, description, ip, country, is_public, start, start_request, num_callback = 0) {
```
Comment (harababurel) [inline client.js]: `start_request` sounds like an action rather than a timestamp. Also, what's the difference between `start` and `start_request`?

File: client.js [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+                        if(data.includes(description)) { // it must also contain the description to be ok
+                            ok = true
+                        } else { // otherwise no
+                            ok = false
+                        }
+                    } else { // if is not the previous 2 cases it means is not correct
+                        ok = false
+                    }
+
+                    var t1 = performance.now()
```
Comment (harababurel) [inline client.js]: Related to the previous comment: the naming for these two variables (`t1` and `start_request`) is inconsistent.

File: client.js [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+                })
+            }
+        )
+        .on("error", (error) => {
+            reject(error.message)
+        })
+    })
+        .then((data) => {
+            if (data[0] === false) {
+                if (performance.now() - start_request >= 300000) { // 5 minutes
```
Comment (harababurel) [inline client.js]: Let's move this into a config.

File: client.js [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+            reject(error.message)
+        })
+    })
+        .then((data) => {
+            if (data[0] === false) {
+                if (performance.now() - start_request >= 300000) { // 5 minutes 
+                    // exceeded the double of the average to find the id of the video
+                    return data
+                }
+                return searchIdAndFeatures(title, id_video, description, ip, country, is_public, start, start_request, num_callback + 1)
```
Comment (harababurel) [inline client.js]: Is `num_callback` used for anything at the moment?

File: client.js [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+            } else {
+                // query name, country, result search, time from start to end of the request, time from start function call to end request
+                if (socket_localhost !== null) {
```
Comment (harababurel) [inline client.js]: Maybe add logging here to have more visibility over what's happening.

File: client.js [hunk 19]
```diff
@@ -0,0 +1,291 @@
...
+                }
+                return data
+            }
+        })
+        .catch((err) => {
+            console.error(err)
+        })
+}
+
+async function searchId(title, id_video, ip, country, start, start_request, num_callback = 0) {
```
Comment (harababurel) [inline client.js]: This function is similar to `searchIdAndFeatures`. Is there a way to reuse some of the code and reduce duplication?

File: client.js [hunk 20]
```diff
@@ -0,0 +1,291 @@
...
+                })
+            }
+        )
+        .on("error", (error) => {
+            reject(error.message)
+        })
+    })
+        .then((data) => {
+            if (data[0] === false) {
+                if (performance.now() - start_request >= 120000) { // 2 minutes
```
Comment (harababurel) [inline client.js]: Hardcoded value

File: client.js [hunk 21]
```diff
@@ -0,0 +1,291 @@
...
+
+if (argv.webserver) {
+    var localServerio = http.createServer(app)
+    var localServer = require("socket.io")(localServerio)
+
+    localServer.on("connection", (socket) => {
+        socket_localhost = socket
+    })
+
+    localServerio.listen(9001, () => {
```
Comment (harababurel) [inline client.js]: It would be nice to have the option of specifying this port via a flag or a config parameter.

File: client.js [hunk 22]
```diff
@@ -0,0 +1,291 @@
...
+        console.log("listening on localhost:9001")
+    })
+
+    // TODO instead of yarn start, would be better to have a react build of it, and run it with express in some port
+    exec("cd client && yarn start")
```
Comment (harababurel) [inline client.js]: This creates the assumption that `yarn` is installed, which is not always the case. It also assumes that the CLI is ran from the root directory of the project.

File: client.js [hunk 23]
```diff
@@ -0,0 +1,291 @@
...
+}
+
+const io = require("socket.io-client")
+const server = io.connect("http://localhost:5000")
```
Comment (harababurel) [inline client.js]: Same point about the port.

File: client.js [hunk 24]
```diff
@@ -0,0 +1,291 @@
...
+
+server.on("connect", () => {
+    if (argv._.includes("upload-basic") || argv._.includes("upload-days")) {
```
Comment (harababurel) [inline client.js]: This means that running a malformed command (e.g. `node client askljd alksjda upload-basic askldjaskld`) will still run the command. Is this ideal, considering possible ambiguity? What happens if the user provides all commands at once?

File: client.js [hunk 25]
```diff
@@ -0,0 +1,291 @@
...
+        console.log(title, id_video)
+
+        if (argv._.includes("upload-basic")) {
+            basicSearch(title, id_video)
+        } else if (argv._.includes("upload-days")) {
+            mutilpleDaySearch(title, id_video)
+        }
+    })
+
+    server.on("upload-video-and-update-server", async (title, id_video, description, privacyStatus) => {
```
Comment (harababurel) [inline client.js]: nit: inconsistent variable case (underscores vs camelCase). Please check [6.2 Rules by identifier type](https://google.github.io/styleguide/jsguide.html#naming-rules-by-identifier-type) for proper naming rules.

File: client.js [hunk 26]
```diff
@@ -0,0 +1,291 @@
...
+    })
+        .then((data) => {
+            if (data[0] === false) {
+                if (performance.now() - start_request >= 120000) { // 2 minutes 
+                    // exceeded the double of the average to find the id of the video
+                    return data
+                }
+                return searchId(title, id_video, ip, country, start, start_request, num_callback + 1)
+            } else {
+                // query name, country, result search, time from start to end of the request, time from start function call to end request
```
Comment (harababurel) [inline client.js]: Usually methods that accept multiple parameters can be made more transparent using a notation similar to this:

```c++
some_function(/*hostname=*/"localhost", /*port=*/8000, ...)
```

I know python allows setting the parameters explicitly (`some_function(hostname="localhost", ...)`) but I'm not sure about the javascript approach.

**Review by harababurel — CHANGES_REQUESTED** [review]
Added a first pass of comments on the few files that I looked into. Will add more at a later time.

File: client.js (hunk 24 above)
Comment (0x5eba) [reply client.js]: Fixed it, now accept only 1 command

File: client.js (hunk 17 above)
Comment (0x5eba) [reply client.js]: I was using if for testing, I don't think it could be useful, use timeout by ms is better

File: client.js (hunk 13 above)
Comment (0x5eba) [reply client.js]:
```
let ok = false
if(!isPublic && !idFound) ok = true
if(isPublic && idFound && data.includes(description)) ok = true
```
This should be ok

File: client.js (hunk 14 above)
Comment (0x5eba) [reply client.js]: The differenct could be of 100ms from `start` to `start_request` but i guess i can delete data

File: README.md (hunk 4 above)
Comment (0x5eba) [reply README.md]: Yes is from Brian's code, his project was called ytPotato ahaha

File: README.md (hunk 1 above)
Comment (0x5eba) [reply README.md:35]: Yes, done it

File: client.js (hunk 7 above)
Comment (0x5eba) [reply client.js]: I called it `cellName`, i don't have better name ideas

File: server.js [hunk 27]
```diff
@@ -0,0 +1,27 @@
...
+	})
+
+	socket.on("upload-video-and-update", () => {
+		new MeasureUploadAndUpdateDelay().run(socket, headless)
+	})
+
+	socket.on("disconnect", () => {})
+})
+
+server.listen(9001, () => {
```
Comment (harababurel) [inline server.js]: Hardcoded port

File: server/config.js [hunk 28]
```diff
@@ -0,0 +1,26 @@
...
+	studioBaseUrl: "http://studio.youtube.com",
+	puppeteerViewportOptions: { width: 1200, height: 800 },
+
+	// The username and password for the user used in testing.  Currently, only a
+	// single user is used.
+	username: "bsebastientest@gmail.com",
+	password: "FKYCZDkle2",
+
+	// username: "bronngreyjoy.557864@gmail.com",
+	// password: "f74d0c7cc1b420720d2affe2aca2d2b7",
```
Comment (harababurel) [inline server/config.js]: These should not be made public. Let's move them to a separate credentials file and add that to `.gitignore`.

File: server/gen_video.py [hunk 29]
```diff
@@ -0,0 +1,14 @@
...
+import cv2
+
+dim = 100
+img = np.zeros((dim,dim,3),np.uint8)
+for x in range(dim):
+    for y in range(dim):
+        img[x,y] = [random.randint(0,255),random.randint(0,255),random.randint(0,255)]
+
+cv2.imwrite(f"./test_videos/{sys.argv[1]}.png",img)
+os.system(f"ffmpeg -r 1 -i ./test_videos/{sys.argv[1]}.png -vcodec mpeg4 -y ./test_videos/{sys.argv[1]}.mp4")
```
Comment (harababurel) [inline server/gen_video.py]: Please make sure to run every source file through an autoformatter (e.g. `autopep8` for python).

File: client/src/Graphs.js [hunk 30]
```diff
@@ -1,32 +1,76 @@
...
+			graphData: [],
+		}
+	}
+
+	componentDidMount() {
+		this.connectionServer()
+	}
+
+	connectionServer() {
+		socket = io.connect("http://localhost:9002", { secure: true })
```
Comment (harababurel) [inline client/src/Graphs.js]: Hardcoded port

**Review by harababurel — CHANGES_REQUESTED** [review]

File: install_dep.sh [hunk 31]
```diff
@@ -0,0 +1,23 @@
+# check that all the dependecies are installed
```
Comment (harababurel) [inline install_dep.sh:4]: This should start with `#!/bin/bash` or `#!/bin/sh`. Also consider adding `set -ex` to print the commands and exit when any of them fails.

File: install_dep.sh [hunk 32]
```diff
@@ -0,0 +1,23 @@
...
+    echo -e "ffmpeg not found, install it before continuing. If you are using ubuntu should be: sudo apt-get install ffmpeg"
+fi
+
+# dependencies for npm
+echo 'Installing dependencies for npm'
+npm install && cd server && npm install && cd ../client && npm install && cd ../
+
+# dependencies for python
+echo 'Installing dependencies for python'
+pip install numpy opencv-python
```
Comment (harababurel) [inline install_dep.sh]: pip might not be installed

File: server/gen_video.py [hunk 33]
```diff
@@ -21,8 +21,14 @@
 import random
 import numpy as np
 import cv2
+from subprocess import call, PIPE
+
+dir_path = os.path.dirname(os.path.realpath(__file__))
+if "test_videos" not in dir_path:
+	dir_path += "/test_videos/"
```
Comment (harababurel) [inline server/gen_video.py]: os.path.join

File: server/gen_video.py [hunk 34]
```diff
@@ -21,8 +21,14 @@
...
+
+call(f"mkdir {dir_path}".split())
```
Comment (harababurel) [inline server/gen_video.py:31]: Consider adding logs to make the behavior more transparent. Same for the other steps below.

**Review by harababurel — APPROVED** [review]

File: config.json [hunk 35]
```diff
@@ -0,0 +1,9 @@
+{
+    "SERVER_PORT": 9001,
+    "CLIENT_WEBSERVER_PORT": 9002,
+    "SEARCH_ID_TIMEOUT": 120000,
+    "SEARCH_ID_AND_FEATURES_TIMEOUT": 300000,
+    "DAYS_OF_SEARCH": 259200000,
+    "INTERVAL_SEARCH_DAYS": 600000,
```
Comment (harababurel) [inline config.json]: What do these values mean? For example, one of them can be interpreted as "Search for 259200000 days" (?)

File: install_dep.sh (hunk 31 above)
Comment (0x5eba) [reply install_dep.sh:4]: Where should I add `set -ex`?

File: config.json (hunk 35 above)
Comment (0x5eba) [reply config.json]: I can't add comments in a json file, should i change it to a .js file?

Comment (0x5eba) [reply config.json]: I will add `_IN_MILLISECONDS` and should be fine

File: install_dep.sh (hunk 31 above)
Comment (harababurel) [reply install_dep.sh:4]: It usually goes on the second line, after `#!/bin/bash`. Keep in mind that I'm not saying this is the right choice, I'm just providing an option.

File: config.json (hunk 35 above)
Comment (harababurel) [reply config.json]: Ideally the field names should be descriptive enough not to require comments.

File: install_dep.sh (hunk 31 above)
Comment (0x5eba) [reply install_dep.sh:4]: Seems good to me
